### PR TITLE
CORDA-3959: Eliminate InternalCordaRPCOps

### PR DESCRIPTION
**Pull request by vkolomeyko**
This change represents a mix of changes:
1. Some files been ported from CE in which case I will clearly identify and marked them.
2. Same or different files been modified in OS to accommodate:
   a. Multi-RPC support and corresponding permissions
   b. Eliminate `InternalCordaRPCOps`

This change will be in a way breaking because it completely removes `InternalCordaRPCOps` from the codebase. This means that if an old RPC client tries to make an RPC call using `InternalCordaRPCOps` interface - this call is not going to work. I do believe this is acceptable as `InternalCordaRPCOps` was never meant for public use and those who use it internally would have to make changes to their code.
Corda Shell is still working with the same commands as before.

After this PR is merged into OS, more work will be needed to bring this change over to ENT. I.e. separate ENT PR to follow.

Suggested place to start code review: https://github.com/corda/corda/pull/6600/files#diff-2ecd7b120b2ed0a7019f6d8a830d0cc5R408

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 1]
```diff
@@ -401,21 +404,28 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
...
+        }.also { it.closeOnStop() })
+
+        val checkpointRPCOpsImpl = Pair(CheckpointRPCOps::class.java, CheckpointRPCOpsImpl(checkpointDumper))
+
+        val attachmentTrustInfoRPCOps = Pair(AttachmentTrustInfoRPCOps::class.java, AttachmentTrustInfoRPCOpsImpl(services.attachmentTrustCalculator))
+
+        return listOf(cordaRPCOpsImpl, checkpointRPCOpsImpl, attachmentTrustInfoRPCOps).map { rpcOpsImplPair ->
+            // Mind that order of proxies is important
+            val targetInterface = rpcOpsImplPair.first
+            val stage1Proxy = AuthenticatedRpcOpsProxy.proxy(rpcOpsImplPair.second, targetInterface)
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:415]: Similar piece of code can also be found in ENT.

File: node/src/main/kotlin/net/corda/node/internal/rpc/proxies/AuthenticatedRpcOpsProxy.kt [hunk 2]
```diff
@@ -1,55 +1,72 @@
...
-     *
-     * TODO: Why is this logic duplicated vs the actual implementation?
-     */
-    override val protocolVersion: Int get() = delegate.nodeInfo().platformVersion
-
-    // Need overriding to pass additional `listOf(logicType)` argument for polymorphic `startFlow` permissions.
-    override fun <T> startFlowDynamic(logicType: Class<out FlowLogic<T>>, vararg args: Any?) = guard("startFlowDynamic", listOf(logicType), ::rpcContext) {
-        delegate.startFlowDynamic(logicType, *args)
-    }
+internal object AuthenticatedRpcOpsProxy {
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/internal/rpc/proxies/AuthenticatedRpcOpsProxy.kt:15]: This code was brought from ENT.

File: node/src/main/kotlin/net/corda/node/internal/rpc/proxies/ThreadContextAdjustingRpcOpsProxy.kt [hunk 3]
```diff
@@ -1,30 +1,28 @@
...
  * without sensible fallbacks to the classloader of the current instance.
  * If clients' CorDapps use one of these libraries, this temporary adjustment can ensure that any provided classes from these libraries will be available during RPC calls.
  */
-internal class ThreadContextAdjustingRpcOpsProxy(private val delegate: InternalCordaRPCOps, private val classLoader: ClassLoader) : InternalCordaRPCOps by proxy(delegate, classLoader) {
-    private companion object {
-        private fun proxy(delegate: InternalCordaRPCOps, classLoader: ClassLoader): InternalCordaRPCOps {
-            val handler = ThreadContextAdjustingRpcOpsProxy.ThreadContextAdjustingInvocationHandler(delegate, classLoader)
-            return Proxy.newProxyInstance(delegate::class.java.classLoader, arrayOf(InternalCordaRPCOps::class.java), handler) as InternalCordaRPCOps
-        }
+internal object ThreadContextAdjustingRpcOpsProxy {
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/internal/rpc/proxies/ThreadContextAdjustingRpcOpsProxy.kt:15]: This code was brought from ENT.

File: node/src/main/kotlin/net/corda/node/internal/security/RPCPermissionResolver.kt [hunk 4]
```diff
@@ -0,0 +1,175 @@
+package net.corda.node.internal.security
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/internal/security/RPCPermissionResolver.kt:1]: This code was brought from ENT.

File: node/src/test/kotlin/net/corda/node/internal/security/RPCPermissionResolverTest.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package net.corda.node.internal.security
```
Comment (vkolomeyko) [inline node/src/test/kotlin/net/corda/node/internal/security/RPCPermissionResolverTest.kt:1]: This code was brought from ENT.

File: tools/shell/src/main/java/net/corda/tools/shell/SshAuthInfo.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package net.corda.tools.shell;
+
+import net.corda.core.messaging.RPCOps;
+import org.crsh.auth.AuthInfo;
+
+public interface SshAuthInfo extends AuthInfo {
```
Comment (vkolomeyko) [inline tools/shell/src/main/java/net/corda/tools/shell/SshAuthInfo.java:6]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaAuthenticationPlugin.kt [hunk 7]
```diff
@@ -1,14 +1,13 @@
...
-import net.corda.core.internal.messaging.InternalCordaRPCOps
+import net.corda.core.messaging.CordaRPCOps
 import net.corda.core.utilities.loggerFor
 import org.apache.activemq.artemis.api.core.ActiveMQSecurityException
 import org.crsh.auth.AuthInfo
 import org.crsh.auth.AuthenticationPlugin
 import org.crsh.plugin.CRaSHPlugin
 
-class CordaAuthenticationPlugin(private val makeRPCConnection: (username: String, credential: String) -> CordaRPCConnection) : CRaSHPlugin<AuthenticationPlugin<String>>(), AuthenticationPlugin<String> {
+internal class CordaAuthenticationPlugin(private val rpcOpsProducer: RPCOpsProducer) : CRaSHPlugin<AuthenticationPlugin<String>>(), AuthenticationPlugin<String> {
```
Comment (vkolomeyko) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/CordaAuthenticationPlugin.kt:10]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaRpcOpsShellCommand.kt [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package net.corda.tools.shell
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.type.TypeFactory
+import net.corda.core.messaging.CordaRPCOps
+
+internal abstract class CordaRpcOpsShellCommand : InteractiveShellCommand<CordaRPCOps>() {
```
Comment (vkolomeyko) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/CordaRpcOpsShellCommand.kt:7]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt [hunk 9]
```diff
@@ -1,17 +1,53 @@
...
+import net.corda.core.internal.utilities.InvocationHandlerTemplate
+import net.corda.core.messaging.RPCOps
 import net.corda.tools.shell.utlities.ANSIProgressRenderer
-import org.crsh.auth.AuthInfo
+import java.lang.reflect.Proxy
 
-class CordaSSHAuthInfo(val successful: Boolean, val rpcOps: InternalCordaRPCOps, val ansiProgressRenderer: ANSIProgressRenderer? = null,
-                       val isSsh: Boolean = false, val rpcConn: CordaRPCConnection? = null) : AuthInfo {
-    override fun isSuccessful(): Boolean = successful
+internal class CordaSSHAuthInfo(private val rpcOpsProducer: RPCOpsProducer,
```
Comment (vkolomeyko) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt:13]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShellCommand.kt [hunk 10]
```diff
@@ -1,23 +1,24 @@
...
-open class InteractiveShellCommand : BaseCommand() {
-    fun ops() = ((context.session as CRaSHSession).authInfo as CordaSSHAuthInfo).rpcOps
-    fun ansiProgressRenderer() = ((context.session as CRaSHSession).authInfo as CordaSSHAuthInfo).ansiProgressRenderer
-    fun objectMapper(classLoader: ClassLoader?): ObjectMapper {
-        val om = ((context.session as CRaSHSession).authInfo as CordaSSHAuthInfo).yamlInputMapper
-        if (classLoader != null) {
-            om.typeFactory = TypeFactory.defaultInstance().withClassLoader(classLoader)
-        }
-        return om
+internal abstract class InteractiveShellCommand<T : RPCOps> : BaseCommand() {
```
Comment (vkolomeyko) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/InteractiveShellCommand.kt:10]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/RPCOpsProducer.kt [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+import net.corda.client.rpc.CordaRPCClient
+import net.corda.client.rpc.CordaRPCClientConfiguration
+import net.corda.client.rpc.GracefulReconnect
+import net.corda.client.rpc.RPCConnection
+import net.corda.client.rpc.internal.RPCClient
+import net.corda.core.messaging.CordaRPCOps
+import net.corda.core.messaging.RPCOps
+
+internal interface RPCOpsProducer {
```
Comment (vkolomeyko) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/RPCOpsProducer.kt:11]: This code was brought from ENT.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt [hunk 12]
```diff
@@ -1,17 +1,53 @@
...
+                                val isSsh: Boolean = false) : SshAuthInfo  {
+    override fun isSuccessful(): Boolean = true
 
-    val yamlInputMapper: ObjectMapper by lazy {
-        createYamlInputMapper(rpcOps)
+    /**
+     * It is necessary to have a cache to prevent creation of too many proxies for the same class. Proxy ensures that RPC connections gracefully
+     * closed when cache entry is eliminated
+     */
+    private val proxiesCache = Caffeine.newBuilder()
```
Comment (rick-r3) [inline tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt:22]: I know this code is from ENT, but.... This cache is limited to the session isn't it?  i.e. per session.  So we don't need to worry about it actually becoming full ?  Otherwise what happens when the 11th one is created and the 1st one is automatically /abruptly closed ?

File: node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt [hunk 13]
```diff
@@ -177,10 +176,10 @@ class RPCServer(
                 val groupedMethods = with(interfaceClass) {
                     /*
                      * Until version 4.3, rpc calls did not include class names.
-                     * Up to this version, only CordaRPCOps and InternalCordaRPCOps were supported.
+                     * Up to this version, only CordaRPCOps was supported.
                      * So, for these classes methods are registered without their class name as well to preserve backwards compatibility.
                      */
-                    if(interfaceClass == CordaRPCOps::class.java || interfaceClass == InternalCordaRPCOps::class.java) {
```
Comment (dimosr) [inline node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt:183]: Is this going to break backwards compatibility (old clients talking to new servers)? Are we OK with it? Don't we need to redirect `InternalCordaRPCOps` commands to the new APIs?

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt (hunk 12 above)
Comment (mnesbit) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt:22]: Yes I don't know what happens to observables when we redo an entry?

File: node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package net.corda.node.internal.checkpoints
+
+import net.corda.core.internal.PLATFORM_VERSION
+import net.corda.core.internal.messaging.CheckpointRPCOps
+import net.corda.node.services.rpc.CheckpointDumperImpl
+
+/**
+ * Implementation of [CheckpointRPCOps]
+ */
+internal class CheckpointRPCOpsImpl(private val checkpointDumper: CheckpointDumperImpl) : CheckpointRPCOps {
```
Comment (mnesbit) [inline node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt]: My only question is if we should widen scope of this to be a bit more related to the Flow management work? i.e. perhaps be a FlowManagerRPCOpsImpl rather than just about checkpoint dumping? Talk to @lankydan and @roastario if there are any sensible OS supported bits for such a thing, or if this is out of scope.

**Review by mnesbit — COMMENTED** [review]
I agree the cache bit may need thinking about/simplifying for OS. Plus wondering if we can do a bit better than a single method RPC interface for checkpoints

**Review by dimosr — COMMENTED** [review]
I had a look with a focus on the rpc parts (not extremely familiar with the shell) and it looks good to me. Left only one comment around backwards compatibility.

File: tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt (hunk 12 above)
Comment (vkolomeyko) [reply tools/shell/src/main/kotlin/net/corda/tools/shell/CordaSSHAuthInfo.kt:22]: This cache is indeed per session.

The cache is keyed by `RPCOps` class name and at the moment we do not even have 10 distinct `RPCOps` - so the cache will never be full.
When we do have more than 10 `RPCOps` interfaces and even if shell commands used within a single session touch all those 10 interfaces there is a chance that we start closing RPC connections. But even then, many shell commands do not do anything with observables, i.e. just like `run nodeInfo()` there are more request/reply like.
So chances having any sort of negative impact are rather remote I would say.

The cache was mainly introduced to re-use RPC connections for previously used `RPCOps` rather than creating a fresh connection every time, which might be costly.

File: node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt (hunk 13 above)
Comment (vkolomeyko) [reply node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt:183]: Yes, @dimosr - it is a valid concern and I put a note in the description of this PR.

I wanted to eliminate all the references to `InternalCordaRPCOps` and instead of translating old calls to new API and translating permission old permission strings to the new FQN of the methods, went for more radical approach.

Should we do it differently, we might need to carry this legacy load in the source code for quite some time and then still there is a risk that some people will not migrate to the new API.
`InternalCordaRPCOps` was always an internal interface and any client using it was doing so in their own risk.

File: node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt (hunk 14 above)
Comment (vkolomeyko) [reply node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt]: @lankydan and @roastario - would you be able to opine on your preference for naming in OS?

File: node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt (hunk 13 above)
Comment (mnesbit) [reply node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt:183]: Only shell talks to InternalCordaRPCOps so I don't think there are any public clients. The one exception is out of process shell and I am fine with saying that this needs to be version matched.

Comment (dimosr) [reply node/src/main/kotlin/net/corda/node/services/rpc/RPCServer.kt:183]: True, I didn't initially think these APIs are used only by the shell.

File: node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt (hunk 14 above)
Comment (roastario) [reply node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt]: so unfortunately all the components which expose the FlowStatus functionality are ENT only. So the OS version would not be able to expose any of this.

Comment (roastario) [reply node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt]: Naming wise, I don't see a harm in calling it something a bit more generic such as FlowManagerRPCOpsImpl, and only adding functionality in ENT. This also would allow us to gradually add functionality to this interface.

Comment (vkolomeyko) [reply node/src/main/kotlin/net/corda/node/internal/checkpoints/CheckpointRPCOpsImpl.kt]: 👍

**Review by dimosr — APPROVED** [review]
LGTM, but expecting approval from the other reviewers with comments too.

**Review by mnesbit — APPROVED** [review]
OK lets do this
